versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java

```diff
@@ -311,6 +331,179 @@ public ReferenceDeletedResult delete(NamedRef namedRef, Hash hash)
+    }
+
+    // Hard coded limit of 100 commits to check. This function will check at max 1000 commits in the
+    // commit log, or until the configured recorded reference-history interval.
+    int toScan = Math.min(headCommitsToScan, 1000);
+    long nowMicros = persist.config().currentTimeMicros();
+    long inconsistencyWindow =
+        TimeUnit.SECONDS.toMicros(persist.config().referencePreviousHeadTimeSpanSeconds());
+
+    CommitObj headCommit;
+    try {
+      headCommit = commitLogic.fetchCommit(hashToObjId(current.pointer()));
+    } catch (ObjNotFoundException e) {
+      return CommitConsistency.COMMIT_INCONSISTENT;
+    }
+    if (headCommit == null) {
+      return CommitConsistency.COMMIT_CONSISTENT;
+    }
+
+    // Start with the HEAD's direct parent, the HEAD itself is checked elsewhere.
+    ObjId commitId = headCommit.directParent();
+
+    // The "initial" result of the commit log cannot be better than the state of the HEAD, so start
+    // with the HEAD's state.
+    CommitConsistency result = current.commitConsistency();
+
+    while (toScan-- > 0) {
+      if (commitId.equals(EMPTY_OBJ_ID)) {
+        break;
+      }
+
+      CommitObj currentCommit;
+
+      try {
+        currentCommit = commitLogic.fetchCommit(commitId);
+        if (currentCommit == null) {
+          break;
+        }
+      } catch (ObjNotFoundException e) {
+        return CommitConsistency.COMMIT_INCONSISTENT;
+      }
+
+      if (nowMicros - currentCommit.created() > inconsistencyWindow) {
+        break;
+      }
+
+      ReferenceHistory.ReferenceHistoryElement elem =
+          checkCommit(commitLogic, indexesLogic, commitId, checked);
+      if (elem.commitConsistency().ordinal() > result.ordinal()) {
+        // State of the checked commit is worse than the current state.
+        result = elem.commitConsistency();
+      }
+      for (ObjId secondaryParent : currentCommit.secondaryParents()) {
+        elem = checkCommit(commitLogic, indexesLogic, secondaryParent, checked);
+        if (elem.commitConsistency().ordinal() > result.ordinal()) {
+          // State of the checked commit is worse than the current state.
+          result = elem.commitConsistency();
+        }
+      }
+
+      if (result == CommitConsistency.COMMIT_INCONSISTENT) {
+        // This is the worst state, can return immediately.
+        return result;
+      }
+
+      commitId = currentCommit.directParent();
+    }
+
+    return result;
+  }
+
+  private ReferenceHistory.ReferenceHistoryElement checkCommit(
+      CommitLogic commitLogic, IndexesLogic indexesLogic, ObjId commit, Set<ObjId> checked) {
+    Hash commitId = objIdToHash(commit);
+    CommitMeta meta = null;
+
+    Set<ObjId> test = new HashSet<>();
+    CommitObj commitObj;
+    try {
+      commitObj = commitLogic.fetchCommit(commit);
+      if (commitObj == null) {
+        return referenceHistoryElement(commitId, CommitConsistency.COMMIT_CONSISTENT, null);
```

Comment:
Why flag it as `COMMIT_CONSISTENT` when the actual commit object is not found?

Reply:
null == no-anchestor-hash